File: logging_test.go
package tea

import (
	"bytes"
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"
	"testing"
)

func TestLogToFile(t *testing.T) {
	path := filepath.Join(t.TempDir(), "log.txt")
	prefix := "logprefix"
	f, err := LogToFile(path, prefix)
	if err != nil {
		t.Error(err)
	}
	log.SetFlags(log.Lmsgprefix)
	log.Println("some test log")
	if err := f.Close(); err != nil {
		t.Error(err)
	}
	out, err := os.ReadFile(path)
	if err != nil {
		t.Error(err)
	}
	if string(out) != prefix+" some test log\n" {
		t.Fatalf("wrong log msg: %q", string(out))
	}
}

type multiWriterLogger struct {
	writers []io.Writer
	prefix  string
	extra   io.Writer
}

func newMultiWriterLogger(extra io.Writer) *multiWriterLogger {
	return &multiWriterLogger{extra: extra}
}

func (l *multiWriterLogger) SetOutput(w io.Writer) {
	l.writers = []io.Writer{w}
	if l.extra != nil {
		l.writers = append(l.writers, l.extra)
	}
}

func (l *multiWriterLogger) SetPrefix(prefix string) {
	l.prefix = prefix
}

func (l *multiWriterLogger) Println(msg string) error {
	if len(l.writers) == 0 {
		return fmt.Errorf("output not configured")
	}
	line := l.prefix + msg + "\n"
	for _, writer := range l.writers {
		if _, err := io.WriteString(writer, line); err != nil {
			return err
		}
	}
	return nil
}

func TestLogToFileWithMultiWriterLogger(t *testing.T) {
	path := filepath.Join(t.TempDir(), "multi-writer-log.txt")
	var buffer bytes.Buffer
	logger := newMultiWriterLogger(&buffer)
	f, err := LogToFileWith(path, "multi", logger)
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() {
		if cerr := f.Close(); cerr != nil {
			t.Error(cerr)
		}
	})
	if err := logger.Println("structured log entry"); err != nil {
		t.Fatal(err)
	}
	contents, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	if string(contents) != "multi structured log entry\n" {
		t.Fatalf("file writer mismatch: %q", string(contents))
	}
	if buffer.String() != "multi structured log entry\n" {
		t.Fatalf("fan-out writer mismatch: %q", buffer.String())
	}
}

func TestLogToFileAppendsExistingContents(t *testing.T) {
	path := filepath.Join(t.TempDir(), "append-log.txt")
	if err := os.WriteFile(path, []byte("legacy entry\n"), 0o600); err != nil {
		t.Fatalf("failed to seed log file: %v", err)
	}
	logger := log.New(io.Discard, "", log.Lmsgprefix)
	f, err := LogToFileWith(path, "append", logger)
	if err != nil {
		t.Fatal(err)
	}
	closed := false
	t.Cleanup(func() {
		if closed {
			return
		}
		if cerr := f.Close(); cerr != nil {
			t.Error(cerr)
		}
	})
	logger.Println("new entry")
	if err := f.Close(); err != nil {
		t.Fatal(err)
	}
	closed = true
	contents, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	const want = "legacy entry\nappend new entry\n"
	if string(contents) != want {
		t.Fatalf("log file mismatch: %q", string(contents))
	}
}

func TestLogToFileSupportsMultipleWriters(t *testing.T) {
	path := filepath.Join(t.TempDir(), "shared-log.txt")
	logger1 := log.New(io.Discard, "", log.Lmsgprefix)
	logger2 := log.New(io.Discard, "", log.Lmsgprefix)
	f1, err := LogToFileWith(path, "proc1", logger1)
	if err != nil {
		t.Fatal(err)
	}
	f2, err := LogToFileWith(path, "proc2", logger2)
	if err != nil {
		t.Fatal(err)
	}
	closed1 := false
	closed2 := false
	t.Cleanup(func() {
		if !closed1 {
			if cerr := f1.Close(); cerr != nil {
				t.Error(cerr)
			}
		}
		if !closed2 {
			if cerr := f2.Close(); cerr != nil {
				t.Error(cerr)
			}
		}
	})
	logger1.Println("first entry")
	logger2.Println("second entry")
	if err := f1.Close(); err != nil {
		t.Fatal(err)
	}
	closed1 = true
	if err := f2.Close(); err != nil {
		t.Fatal(err)
	}
	closed2 = true
	contents, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	const want = "proc1 first entry\nproc2 second entry\n"
	if string(contents) != want {
		t.Fatalf("unexpected log aggregation: %q", string(contents))
	}
}

func TestLogToFileWithMultiWriterLoggerToPipe(t *testing.T) {
	path := filepath.Join(t.TempDir(), "stderr-fanout.txt")
	pr, pw, err := os.Pipe()
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() {
		pr.Close()
		pw.Close()
	})
	logger := newMultiWriterLogger(pw)
	f, err := LogToFileWith(path, "stderr", logger)
	if err != nil {
		t.Fatal(err)
	}
	closed := false
	t.Cleanup(func() {
		if closed {
			return
		}
		if cerr := f.Close(); cerr != nil {
			t.Error(cerr)
		}
	})
	if err := logger.Println("mirrored log entry"); err != nil {
		t.Fatal(err)
	}
	if err := f.Close(); err != nil {
		t.Fatal(err)
	}
	closed = true
	if err := pw.Close(); err != nil {
		t.Fatal(err)
	}
	contents, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	pipeContents, err := io.ReadAll(pr)
	if err != nil {
		t.Fatal(err)
	}
	const want = "stderr mirrored log entry\n"
	if string(contents) != want {
		t.Fatalf("file writer mismatch: %q", string(contents))
	}
	if string(pipeContents) != want {
		t.Fatalf("pipe writer mismatch: %q", string(pipeContents))
	}
}

File: packages/tests/src/logging/logging.test.ts
import { mkdtempSync, readFileSync, writeFileSync } from 'node:fs';
import type { WriteStream } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { PassThrough } from 'node:stream';
import { describe, expect, it } from 'vitest';

import { LogOptionsSetter, LogToFile, createMultiWriterLogOptions } from '@bubbletea/tea';

class TestLogger implements LogOptionsSetter {
  private output: NodeJS.WritableStream | null = null;
  private prefix = '';

  setOutput(output: NodeJS.WritableStream): void {
    this.output = output;
  }

  setPrefix(prefix: string): void {
    this.prefix = prefix;
  }

  log(message: string): void {
    if (!this.output) {
      throw new Error('output not configured');
    }
    this.output.write(`${this.prefix}${message}\n`);
  }
}

class MultiWriterLogger implements LogOptionsSetter {
  private output: NodeJS.WritableStream | null = null;
  private prefix = '';

  constructor(private readonly extra: NodeJS.WritableStream) {}

  setOutput(output: NodeJS.WritableStream): void {
    this.output = output;
  }

  setPrefix(prefix: string): void {
    this.prefix = prefix;
  }

  log(message: string): void {
    if (!this.output) {
      throw new Error('output not configured');
    }
    const line = `${this.prefix}${message}\n`;
    this.output.write(line);
    this.extra.write(line);
  }
}

const createTempLogPath = (): string => {
  const dir = mkdtempSync(join(tmpdir(), 'bubbletea-log-'));
  return join(dir, 'log.txt');
};

const closeStream = (stream: WriteStream): Promise<void> =>
  new Promise((resolve, reject) => {
    stream.end((err) => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });

describe('LogToFile (logging_test.go parity)', () => {
  it('writes prefix and log entry to the configured file', async () => {
    const logger = new TestLogger();
    const path = createTempLogPath();
    const prefix = 'logprefix';

    const stream = LogToFile(path, prefix, logger);
    logger.log('some test log');
    await closeStream(stream);

    const contents = readFileSync(path, 'utf8');
    expect(contents).toBe('logprefix some test log\n');
  });

  it('supports custom loggers that fan out to multiple writers', async () => {
    const extra = new PassThrough();
    const seen: string[] = [];
    extra.on('data', (chunk) => seen.push(chunk.toString('utf8')));

    const logger = new MultiWriterLogger(extra);
    const path = createTempLogPath();
    const stream = LogToFile(path, 'multi', logger);

    logger.log('structured log entry');
    await closeStream(stream);

    const contents = readFileSync(path, 'utf8');
    expect(contents).toBe('multi structured log entry\n');
    expect(seen.join('')).toBe('multi structured log entry\n');
  });

  it('appends to existing files without truncating prior entries', async () => {
    const originals = {
      log: console.log,
      info: console.info,
      warn: console.warn,
      error: console.error
    } as const;

    const path = createTempLogPath();
    writeFileSync(path, 'legacy entry\n', 'utf8');
    const stream = LogToFile(path, 'append');

    try {
      console.log('new entry');
      await closeStream(stream);
      expect(console.log).toBe(originals.log);
      expect(console.info).toBe(originals.info);
      expect(console.warn).toBe(originals.warn);
      expect(console.error).toBe(originals.error);
    } finally {
      console.log = originals.log;
      console.info = originals.info;
      console.warn = originals.warn;
      console.error = originals.error;
    }

    const contents = readFileSync(path, 'utf8');
    expect(contents).toBe('legacy entry\nappend new entry\n');
  });

  it('allows sequential processes to append to the same log file', async () => {
    const originals = {
      log: console.log,
      info: console.info,
      warn: console.warn,
      error: console.error
    } as const;

    const path = createTempLogPath();
    const first = LogToFile(path, 'proc1');
    try {
      console.log('first entry');
      await closeStream(first);
      const second = LogToFile(path, 'proc2');
      console.log('second entry');
      await closeStream(second);
      expect(console.log).toBe(originals.log);
      expect(console.info).toBe(originals.info);
      expect(console.warn).toBe(originals.warn);
      expect(console.error).toBe(originals.error);
    } finally {
      console.log = originals.log;
      console.info = originals.info;
      console.warn = originals.warn;
      console.error = originals.error;
    }

    const contents = readFileSync(path, 'utf8');
    expect(contents).toBe('proc1 first entry\nproc2 second entry\n');
  });
});

describe('createMultiWriterLogOptions', () => {
  it('fans out console output to additional streams', async () => {
    const extra = new PassThrough();
    const seen: string[] = [];
    extra.on('data', (chunk) => seen.push(chunk.toString('utf8')));

    const path = createTempLogPath();
    const stream = LogToFile(path, 'multi', createMultiWriterLogOptions([extra]));

    console.log('helper log entry');
    await closeStream(stream);

    const contents = readFileSync(path, 'utf8');
    expect(contents).toBe('multi helper log entry\n');
    expect(seen.join('')).toBe('multi helper log entry\n');
  });

  it('mirrors logs to stderr-like streams without closing them', async () => {
    const originals = {
      log: console.log,
      info: console.info,
      warn: console.warn,
      error: console.error
    } as const;

    const extra = new PassThrough();
    const seen: string[] = [];
    let closed = false;
    extra.on('data', (chunk) => seen.push(chunk.toString('utf8')));
    extra.once('close', () => {
      closed = true;
    });
    extra.once('finish', () => {
      closed = true;
    });

    const path = createTempLogPath();
    const stream = LogToFile(path, 'stderr', createMultiWriterLogOptions([extra]));

    try {
      console.error('mirrored log entry');
      await closeStream(stream);
      expect(console.log).toBe(originals.log);
      expect(console.info).toBe(originals.info);
      expect(console.warn).toBe(originals.warn);
      expect(console.error).toBe(originals.error);
    } finally {
      console.log = originals.log;
      console.info = originals.info;
      console.warn = originals.warn;
      console.error = originals.error;
      extra.destroy();
    }

    const contents = readFileSync(path, 'utf8');
    expect(contents).toBe('stderr mirrored log entry\n');
    expect(seen.join('')).toBe('stderr mirrored log entry\n');
    expect(closed).toBe(false);
  });
});

describe('Console patching lifecycle', () => {
  it('restores console methods once the log stream closes', async () => {
    const originals = {
      log: console.log,
      info: console.info,
      warn: console.warn,
      error: console.error
    } as const;

    const path = createTempLogPath();
    const stream = LogToFile(path, 'scoped');

    expect(console.log).not.toBe(originals.log);
    expect(console.info).not.toBe(originals.info);
    expect(console.warn).not.toBe(originals.warn);
    expect(console.error).not.toBe(originals.error);

    console.log('scoped entry');

    try {
      await closeStream(stream);
      expect(console.log).toBe(originals.log);
      expect(console.info).toBe(originals.info);
      expect(console.warn).toBe(originals.warn);
      expect(console.error).toBe(originals.error);
    } finally {
      console.log = originals.log;
      console.info = originals.info;
      console.warn = originals.warn;
      console.error = originals.error;
    }

    const contents = readFileSync(path, 'utf8');
    expect(contents).toBe('scoped scoped entry\n');
  });
});
